src/clonemap.rs

```diff
@@ -0,0 +1,717 @@
+    /// observes an empty slot and manages to insert the new element).
+    pub fn get_or_insert_with<F>(&self, key: K, create: F) -> ExistingOrNew<(K, V)>
+    where
+        F: FnOnce() -> V,
+    {
+        let pin = crossbeam_epoch::pin();
+
+        self.raw
+            .get_or_insert_with(
+                key,
+                |key| {
+                    let value: V = create();
+                    CloneMapPayload((key, value))
+                },
+                &pin,
+            )
+            .map(|payload| (payload.0).clone())
+    }
+
+    /// Looks up or inserts a default value of an element.
+    ///
+    /// This is like [get_or_insert_with][CloneConMap::get_or_insert_with],
+    /// but a default value is used instead of manually providing a
+    /// closure.
+    pub fn get_or_insert_default(&self, key: K) -> ExistingOrNew<(K, V)>
+    where
+        V: Default,
+    {
+        self.get_or_insert_with(key, V::default)
+    }
+}
+
+impl<K, V, S> CloneConMap<K, V, S>
+where
+    K: Clone + Hash + Eq,
+    V: Clone,
+    S: BuildHasher,
+{
+    /// Creates a new empty map, but with the provided hasher implementation.
+    pub fn with_hasher(hasher: S) -> Self {
+        Self {
+            raw: Raw::with_hasher(hasher),
+        }
+    }
+
+    /// Looks up an element.
+    pub fn get<Q>(&self, key: &Q) -> Option<(K, V)>
+    where
+        Q: ?Sized + Eq + Hash,
+        K: Borrow<Q>,
+    {
+        let pin = crossbeam_epoch::pin();
+        self.raw.get(key, &pin).map(|r| (r.0).clone())
+    }
+
+    /// Removes an element identified by the given key, returning it.
+    pub fn remove<Q>(&self, key: &Q) -> Option<(K, V)>
+    where
+        Q: ?Sized + Eq + Hash,
+        K: Borrow<Q>,
+    {
+        let pin = crossbeam_epoch::pin();
+        self.raw.remove(key, &pin).map(|r| (r.0).clone())
+    }
+}
+
+impl<K, V, S> CloneConMap<K, V, S>
+where
+    K: Clone + Hash + Eq,
+    V: Clone,
+{
+    /// Checks if the map is currently empty.
+    ///
+    /// Note that due to the nature of concurrent map, this is
+    /// inherently racy ‒ another thread may add or remove elements
+    /// between you call this method and act based on the result.
+    pub fn is_empty(&self) -> bool {
+        self.raw.is_empty()
+    }
+
+    /// Returns an iterator through the elements of the map.
+    pub fn iter(&self) -> Iter<K, V, S> {
+        Iter {
+            inner: raw::iterator::Iter::new(&self.raw),
+        }
+    }
+}
+
+impl<K, V> Default for CloneConMap<K, V>
+where
+    K: Clone + Hash + Eq,
+    V: Clone,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V, S> Debug for CloneConMap<K, V, S>
+where
+    K: Debug + Clone + Hash + Eq,
+    V: Debug + Clone,
+{
+    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
+        let mut d = fmt.debug_map();
+        for (k, v) in self {
+            d.entry(&k, &v);
+        }
+        d.finish()
```

Comment:
I wonder if the `fmt.debug_map().entries(self.iter())` would work in this case. It didn't work in case of the `Arc`s, because of lifetimes, but if it worked here, it would be shorter and a bit nicer.

But of course, this one is fine too.

Reply:
You're right, it works.
Just added a silly test to check it.